app/helpers/gtl_helper.rb

```diff
@@ -0,0 +1,125 @@
+  # so that `render_gtl` can be a pure function.
+  #
+  def render_gtl_outer(no_flash_div)
+    parent_id = @report_data_additional_options.try(:[], :parent_id)
+
+    options = {
+      :model_name                     => model_to_report_data,
+      :no_flash_div                   => no_flash_div || false,
+      :gtl_type_string                => @gtl_type,
+      :active_tree                    => (x_active_tree unless params[:display] || @use_action),
+      :parent_id                      => parent_id,
+      :selected_records               => gtl_selected_records,
+
+      :display                        => @display,
+      :sort_col                       => @sortcol,
+      :sort_dir                       => @sortdir,
+      :explorer                       => @explorer,
+      :view                           => @view,
+      :db                             => @db,
+
+      :report_data_additional_options => @report_data_additional_options,
+    }
+
+    render_gtl(options)
+  end
+
+  # This is a pure function. All the generated markup depends only and fully on the `options`.
+  # Assert about this function calls in your controller specs.
+  #
+  def render_gtl(options)
+    capture do
+      concat(render_gtl_markup(options[:no_flash_div]))
+      concat(render_gtl_javascripts(options))
+    end
+  end
+
+  def render_gtl_markup(no_flash_div)
+    content_tag('div', :id => 'miq-gtl-view', "ng-controller" => "reportDataController as dataCtrl") do
+      capture do
+        concat(render(:partial => 'layouts/flash_msg')) unless no_flash_div
+        concat(miq_tile_view)
+        concat(miq_data_table)
+        concat(
+          content_tag(
+            'div',
+            render(:partial => "layouts/info_msg", :locals => {:message => _("No Records Found.")}),
+            :class    => 'no-record',
+            "ng-show" => "!dataCtrl.settings.isLoading && dataCtrl.gtlData.rows.length === 0"
+          )
+        )
+      end
+    end
+  end
+
+  def render_gtl_javascripts(options)
+    parent_id_escaped = (h(j_str(options[:parent_id])) unless options[:display].nil?)
+
+    javascript_tag <<EOJ
+      sendDataWithRx({unsubscribe: 'reportDataController'});
+      miq_bootstrap('#miq-gtl-view', 'ManageIQ.report_data');
+      sendDataWithRx({initController: {
+        name: 'reportDataController',
+        data: {
+          additionalOptions: #{options[:report_data_additional_options].to_json},
+          modelName: '#{h(j_str(options[:model_name]))}',
+          activeTree: '#{options[:active_tree]}',
+          gtlType: '#{h(j_str(options[:gtl_type_string]))}',
+          parentId: '#{parent_id_escaped}',
+          sortColIdx: '#{options[:sort_col]}',
+          sortDir: '#{options[:sort_dir]}',
+          isExplorer: '#{options[:explorer]}' === 'true' ? true : false,
+          records: #{!options[:selected_records].nil? ? h(j_str(options[:selected_records].to_json)) : "\'\'"},
```

Comment:
you can use try here as well to make it a bit shorter.

also, in general, maybe worth extracting all of the options to a methods, so you can test it easily.

Reply:
Really these lines are not the subject of this PR. I'll fix that in another one.